### PR TITLE
Hibernate with Panache : throw an IAE when using a sort on a named query

### LINKED ISSUE #21067: Hibernate Panache sorting not working with Named Query
### Describe the bug

Sorting not working with Named Query

```
@Entity
@NamedQueries({
        @NamedQuery(name = "countFruit",
                query = "FROM Fruit")
})
public class Fruit extends PanacheEntity {

    @Column(length = 40, unique = true)
    public String name;

    public static List<Fruit> listSortedFruitsByNamedQuery() {
        return list("#countFruit", Sort.by("name").ascending(), new HashMap<>()); // Not sorting
    }

    public static List<Fruit> listSortedFruits() {
        return list("FROM Fruit", Sort.by("name").ascending(), new HashMap<>()); //Sorting is done
    }

}
```

### Expected behavior

find, list and stream functions should support sorting for named queries also. 

### Actual behavior

No errors, but the result list is not sorted if named query is used 

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "11.0.11"

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.3.0.FINAL

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3

### Additional information

Possible related issue: https://github.com/quarkusio/quarkus/issues/20758

### PR DESCRIPTION
**Pull request by loicmathieu**
Fixes #21067

### CODE REVIEW

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java [hunk 1]
```diff
@@ -196,8 +196,11 @@ public PanacheQueryType find(Class<?> entityClass, String query, Object... param
     public PanacheQueryType find(Class<?> entityClass, String query, Sort sort, Object... params) {
         String findQuery = PanacheJpaUtil.createFindQuery(entityClass, query, paramCount(params));
         EntityManager em = getEntityManager(entityClass);
-        // FIXME: check for duplicate ORDER BY clause?
         if (PanacheJpaUtil.isNamedQuery(query)) {
+            if (sort != null) {
+                throw new IllegalArgumentException(
+                        "Sort cannot be used with named queries, add an order by clause to you named query instead");
```
Comment (gastaldi) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:202]: Minor typo:
```suggestion
                        "Sort cannot be used with named queries, add an order by clause to your named query instead");
```

But I think we should display the named query used here?

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java [hunk 2]
```diff
@@ -212,8 +215,11 @@ public PanacheQueryType find(Class<?> entityClass, String query, Map<String, Obj
     public PanacheQueryType find(Class<?> entityClass, String query, Sort sort, Map<String, Object> params) {
         String findQuery = PanacheJpaUtil.createFindQuery(entityClass, query, paramCount(params));
         EntityManager em = getEntityManager(entityClass);
-        // FIXME: check for duplicate ORDER BY clause?
         if (PanacheJpaUtil.isNamedQuery(query)) {
+            if (sort != null) {
+                throw new IllegalArgumentException(
+                        "Sort cannot be used with named queries, add an order by clause to you named query instead");
```
Comment (gastaldi) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:221]:
```suggestion
                        "Sort cannot be used with named queries, add an order by clause to your named query instead");
```

**Review by Sanne — DISMISSED** [review]
Awesome, thanks - even added test coverage :+1: 

Consider it approved but would be nice to apply @gastaldi 's suggestions first.

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java (hunk 1 above)
Comment (Sanne) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:202]: right would be nice to call out the named query name

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java (hunk 2 above)
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java]: Thanks, fixed

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java (hunk 1 above)
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java]: Good suggestion! I added it.

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java [hunk 3]
```diff
@@ -196,9 +196,13 @@ public PanacheQueryType find(Class<?> entityClass, String query, Object... param
     public PanacheQueryType find(Class<?> entityClass, String query, Sort sort, Object... params) {
         String findQuery = PanacheJpaUtil.createFindQuery(entityClass, query, paramCount(params));
         EntityManager em = getEntityManager(entityClass);
-        // FIXME: check for duplicate ORDER BY clause?
         if (PanacheJpaUtil.isNamedQuery(query)) {
             String namedQuery = query.substring(1);
+            if (sort != null) {
+                throw new IllegalArgumentException(
+                        "Sort cannot be used with named query, add an order by clause to the named query " + namedQuery
+                                + " instead");
```
Comment (gastaldi) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:204]: That looks better, but I miss some quotes in the error message, maybe also display the entity containing the named query would help?

`You cannot use Sort.by on named queries. Try adding an "order by" clause to the named query "xxx" defined in the entity "yyyy" instead`

Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java:204]: I think about the quotes but decided it's not worth it but I can add them.
Regarding the entity, I have the information but as a named query is unique it's easily findable.
Keep in mind that the issue will arrise at developement time, at this time the developer knew what it just modified

Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java]: Quotes added!

**Review by gastaldi — APPROVED** [review]

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java [hunk 4]
```diff
@@ -966,7 +969,10 @@
                                 return assertThrows(PanacheQueryException.class,
                                         () -> personDao.find("#Person.namedQueryNotFound").list(),
                                         "singleResult should have thrown");
-                            }).flatMap(v -> namedQueryRepository.list("#NamedQueryMappedSuperClass.getAll"))
+                            }).flatMap(v -> assertThrows(IllegalArgumentException.class,
+                                    () -> personDao.list("#Person.getByName", Sort.by("name"), Parameters.with("name", "stef")),
+                                    "sort should hav thrown"))
```
Comment (gastaldi) [inline integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java:974]:
```suggestion
                                    "Should have thrown sort exception"))
```

**Review by gastaldi — APPROVED** [review]

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java (hunk 4 above)
Comment (loicmathieu) [reply integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/TestEndpoint.java]: done

**Comment by gsmet** [comment]
@loicmathieu I let you add a note in the migration guide for 2.7. Thanks.

**Pull request merged**